---
title: "Analysis"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
## Written by Duc Tran at University of Nevada Reno
## Additional notes made by Egle Cekanaviciute at NASA Ames and marked EC. egle.cekanaviciute@nasa.gov

library(gridExtra)
library(ggbio)
library(biovizBase)
library(GenomicRanges)
library(rtracklayer)
library(snow)
library(data.table)
## setwd("~/Space-Radiation-Mouse-GWAS-main") - that was for Duc
## EC directory below
setwd("~/Desktop/Mouse GWAS paper/For Genetics 2022/Analysis") ## this is for ALL timepoints
## setwd("~/Desktop/Mouse GWAS paper/For Genetics 2022/Analysis/Early_timepoints") ## this is for EARLY timepoints 
## setwd("~/Desktop/Mouse GWAS paper/For Genetics 2022/Analysis/Late_timepoints") ## this is for LATE timepoints 

```

##Convert to qtl
```{r}
for (rad in c("Ar", "Fe", "X-ray", "Bgd")) {
  if(!dir.exists(paste0("./qtl/",rad,"/"))) {
    dir.create(paste0("./qtl/",rad,"/"), recursive = T)
    file.copy("./data/gwas.yaml", paste0("./qtl/",rad,"/gwas.yaml"))
  } 
  
  pheno<-read.csv("./data/phenotypes_by_UserID.csv",header=TRUE)
  tmp <- read.csv('./data/BGD_phenotypes_by_UserID_AVGs_20210415.csv')
  tmp <- data.frame(UserID = tmp$UserID, let = "Average_Bgd", hour=4, Strain = tmp$Strain, intercept = 1, Bgd = tmp$average_Bgd, FociPerGy = 0)
  pheno <- rbind(pheno, tmp)
  
  # Prepare result table for all Bgd and FociPerGy phenotypes
  plist<-c("Bgd","FociPerGy")
  letlist<-unique(pheno$let)
  
  ## EC: at this point we can split time into early, late and all
  timelist<-unique(pheno$hour) ## this uses ALL time points: 4, 8, 24, 48 hours
  ## timelist = as.integer(c(4,8)) ## this uses ONLY EARLY time points: 4 and 8 hours
  ## timelist = as.integer(c(24,48)) ## this uses ONLY LATE time points: 24 and 48 hours
  
  cnt = 1 # indexation to keep track of the increment of colnames
  
  gwas_pheno <- data.frame(fpg = NULL)
  
  gwas_phenocovar <- data.frame(pheno = "fpg", measure = "fpg")
  
  gwas_samplecovar <- data.frame(rad = NULL, time = NULL, strain = NULL)
    
  for (i in 1:length(letlist)) { # Screen all LET conditions
    for (j in 1:length(timelist)) { # Screen all time points
      for (n in 1:length(plist)) { # Both Bgd and FociPerGy phenotypes
        if(letlist[i] == "Si 350 MeV/n") next
        if(letlist[i] == "Average_Bgd" & (timelist[j] != 4 | plist[n] == "FociPerGy") ) next
        if(letlist[i] != "Average_Bgd" & plist[n] != "FociPerGy" ) next
        if(!grepl(rad, letlist[i])) next
        
        bgd_pheno <- pheno[pheno$let== letlist[i] & pheno$hour==timelist[j], "Bgd"]
        names(bgd_pheno) <- as.character(pheno[pheno$let== letlist[i] & pheno$hour==timelist[j] , "UserID"])
        
        subpheno<-pheno[pheno$let==letlist[i] & pheno$hour==timelist[j],plist[n]] # Select phenotypes that correspond to the LET and time point parameters
        finalpheno<-as.data.frame(t(subpheno))
        colnames(finalpheno)<-pheno[pheno$let==letlist[i] & pheno$hour==timelist[j],"Strain"]
        
        p<-finalpheno
        
        if(length(unique(names(p))) == 0) next
        
        gwas_pheno <- rbind(gwas_pheno, data.frame(fpg = as.numeric(p)))
      
        tmp <- data.frame(rad = c("Ar", "Fe", "X-ray", "Bgd")[sapply(c("Ar", "Fe", "X-ray", "Bgd"), function(x) grepl(x, letlist[i]))],
                        time = timelist[j],
                        strain = names(p)
                                                            )
        gwas_samplecovar <- rbind(gwas_samplecovar, tmp)
        
        cnt = cnt + 1
      }
    }
  }
  gwas_pheno$id <- 1:nrow(gwas_pheno)
  gwas_pheno <- gwas_pheno[,c(2,1)]

  write.csv(gwas_pheno, file = paste0("./qtl/",rad,"/gwas_pheno.csv"), row.names = F, quote = F)
  write.csv(gwas_phenocovar, file = paste0("./qtl/",rad,"/gwas_phenocovar.csv"), row.names = F, quote = F)
  write.csv(gwas_samplecovar, file = paste0("./qtl/",rad,"/gwas_gwas_samplecovar.csv"), row.names = F, quote = F)
  
  strain_map <- as.character(gwas_samplecovar$strain)
  
  snps<-read.table("./data/Merged_GENO.txt", header=TRUE) 
  
  for (i in 4:ncol(snps)) {
    snps[,i] <- as.character(snps[,i])
  }
  
  snps[snps == "T"] <- "A"
  snps[snps == "C"] <- "G"
  snps[snps == "H" | snps == "N"] <- "-"
  
  colnames(snps)[colnames(snps) == "CH3"] <- "C3H"
  
  snps <- snps[, c(colnames(snps)[1:3], unique(strain_map))]
  snps <- snps[apply(snps, 1, function(x) {
    x <- x[-c(1,2,3)]
    if(all(x == "-")) return(FALSE)
    if(length(unique(x[x!="-"])) < 2) return(FALSE)
    if(min(table(x[x!="-"])) < 4) return(FALSE)
    return(TRUE)
  }), ]
  
  tmp <- cbind(data.frame(id = 1:nrow(gwas_pheno), stringsAsFactors = F), t(snps[, strain_map]))
  colnames(tmp) <- c("id", as.character(snps$marker))
  rownames(tmp) <- NULL
  tmp <- apply(tmp, 2, as.character)
  fwrite(tmp, file = paste0("./qtl/",rad,"/gwas_geno.csv"), row.names = F, quote = F)
  
  tmp <- snps[, 1:3]
  colnames(tmp) <- c("marker", "chr", "pos")
  tmp$pos <- tmp$pos/2e6
  write.csv(tmp, file = paste0("./qtl/",rad,"/gwas_gmap.csv"), row.names = F, quote = F)
}

```

```{r}
library(qtl2)
for (rad in c("Ar", "Fe", "X-ray", "Bgd")) {
  myQTLData <- read_cross2(file = paste0("./qtl/",rad,"/gwas.yaml") )
  gwas_samplecovar <- read.csv(paste0("./qtl/",rad,"/gwas_gwas_samplecovar.csv"))
  
  map <- insert_pseudomarkers(map=myQTLData$gmap, step=1)
  
  pr <- calc_genoprob(cross=myQTLData, map=map, cores=4)
  
  kinship <- calc_kinship(pr, use_allele_probs = F)
  
  covar <- gwas_samplecovar[, "time", drop=F]
  covar$time <- as.numeric(covar$time)
  addcovar <- model.matrix(~time, data = covar)[,-1]
  out <- scan1(pr, myQTLData$pheno, addcovar = addcovar, kinship = kinship, cores=4)
  saveRDS(out, paste0("./qtl/", rad, "/out.rds"))
}
```


## Gene mapping
```{r}
library(biomaRt)
library(GenomicRanges)
library(fastmatch)
library(tidyverse)

# Load reactome gene sets
reactome_pathways <- GSA::GSA.read.gmt('./data/gs_reactome.gmt') 

reactome_genes <- as.numeric(unique(Reduce(union, reactome_pathways$genesets)))
reactome_gene_set_names <- reactome_pathways$geneset.descriptions 
names(reactome_gene_set_names) <- NULL
reactome_gene_set_short_names <- reactome_pathways$geneset.names 
names(reactome_gene_set_short_names) <- NULL

# Mapping chromosome positions to entrezgene_id
# ensembl <- useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl")
# genes_id <- getBM(attributes=c('ensembl_gene_id', 'chromosome_name','start_position','end_position', 'entrezgene_id', 'mgi_symbol'), mart = ensembl)
# genes_id <- genes_id[which(genes_id$entrezgene_id %in% reactome_genes), ] # Keep genes in reactome pathways
# saveRDS(genes_id, "./data/gene_id.rds")

genes_id <- readRDS("./data/gene_id.rds")

# Create GRanges object for reference genome
grl <- GRanges(seqnames = genes_id$chromosome_name, ranges = IRanges(start = genes_id$start_position-25000, end = genes_id$end_position+25000, names = genes_id$entrezid))

# Find overlaps between SNP positions and reference
chr <- snps$chromosome
chr <- str_replace(chr, "M", "MT")
start <- snps$position.b38.
end <- snps$position.b38. 
gr <- GRanges(seqnames = chr, ranges = IRanges(start = start, end = end) )
g_match <- findOverlaps(gr, grl)

mapping <- data.frame(g_match) %>% group_split(subjectHits)
names(mapping) <- genes_id$entrezgene_id[sapply(mapping, function(x) x$subjectHits[1])]
mapping <- lapply(mapping, function(x) x$queryHits)

snps_entrez_id <- names(mapping)

# Keep only genes in reactome pathways
intersect_genes <- intersect(unique(snps_entrez_id), reactome_genes)
intersect_genes <- intersect_genes[!is.na(intersect_genes)]

intersect_genes_symbol <- sapply(intersect_genes, function(x) genes_id$mgi_symbol[which(genes_id$entrezgene_id == x)[1]])


#Filter gene sets, only keep genes in analysis
reactome_pathways$genesets <- lapply(reactome_pathways$genesets, function(x) x[which(as.numeric(x) %in% intersect_genes)])
```

## Gene scoring and pathway analysis
```{r}

fishersMethod <- function(x) {
    pchisq(-2 * sum(log(x)), df=2*length(x), lower=FALSE)
}
stoufferMethod <- function(x) {
    pnorm(sum(qnorm(x)) / sqrt(length(x)))
}
averageCumulative <- function(x) {
    n <- length(x)
    a <- mean(x)
    1/factorial(n) * sum(sapply(0:floor(n*a), 
                function(k) (-1)^k * choose(n,k) * (n*a-k)^(n)))
}
addCLT <- function(x) {
    if(sum(is.na(x))>0) NA
    else {
        n <- length(x)
        if (n <= 20) {
            averageCumulative(x)
        } else {
            pnorm(mean(x),1/2,sqrt(1/(12*n)),lower=TRUE)
        }  
    }
}

calculate_p <- function(mapping, snps_p, use_na = F, method = "addCLT")
{
  res <- rep(NA, length(mapping))
  names(res) <- names(mapping)
  if(use_na) snps_p[is.na(snps_p)] <- 1
  for (id in names(mapping)) {
    idx <- mapping[[id]]
    if(sum(!is.na(snps_p[idx])) > 0) {
      p <- snps_p[idx]
      p <- p[!is.na(p)]
      p <- p*(1-2*1e-10) + 1e-10
      if(method == "sidak") {
        res[id] <- 1 - (1 - min(p))^((length(p) + 1)/2)
      } else if (method == "addCLT") {
        res[id] <- addCLT(p)
      } else if (method == "stouffer") {
        res[id] <- stoufferMethod(p)
      } else if (method == "fisher") {
        res[id] <- fishersMethod(p)
      }
    }
  }
  res
}

run_ORA <- function(gene_sets, gene_set_names, genes_p, seed=1) {
  set.seed(seed)
  DEGenes <- names(genes_p)[genes_p < 0.05]
  
  thres <- 10
  done <- F
  while (!done) {
    idx <- which(sapply(gene_sets, function(gs) {
      if(sum(gs %in% DEGenes) > thres & length(gs) < 500) return(T) else return(F)
    }))
    if (length(idx) > 200) {
      done <- T
    }
    thres <- thres - 1
  }
  
  gene_sets <- lapply(idx, function(i) gene_sets[[i]])
  gene_set_names <- gene_set_names[idx]
  
  res <- lapply(gene_sets, function(gs){
    
    wBallDraw <- intersect(gs, DEGenes) %>%  length() - 1
    if (wBallDraw < 0) return(1)
    
    wBall <- length(DEGenes)
    bBall <- length(genes_p) - length(DEGenes)
    ballDraw <- length(intersect(gs, names(genes_p)))
    
    1 - phyper(wBallDraw, wBall, bBall, ballDraw)
    
  }) %>% unlist() %>% data.frame(stringsAsFactors = F)

  colnames(res) <- 'p.value'
  res$p.adj <- p.adjust(res$p.value, "fdr")
  res$pathway <- gene_set_names
  res <- res %>% tidyr::drop_na()
  res
}

run_FGSEA <- function(gene_sets, gene_set_names, genes_z, genes_p, perm=1e4, seed=1) {
  names(gene_sets) <- gene_set_names
  
  thres <- 10
  done <- F
  while (!done) {
    idx <- which(sapply(gene_sets, function(gs) {
    if(sum(genes_p[gs] < 0.05, na.rm = T) > thres & length(gs) < 500) return(T) else return(F)
    }))
    if (length(idx) > 200) {
      done <- T
    }
    thres <- thres - 1
  }
  
  gene_sets <- lapply(idx, function(i) gene_sets[[i]])
  gene_set_names <- gene_set_names[idx]

  set.seed(seed)
  res <- fgsea(pathways = gene_sets,
               stats = genes_z,
               nperm = perm)
  data.frame(p.value = res$pval, p.adj = p.adjust(res$pval, "fdr"), pathway = as.character(res$pathway))
}

lodToPval <-
  function(x)
{
  pchisq(x*(2*log(10)),df=1,lower.tail=FALSE)/2
}
```


## EC - making Manhattan plots
```{r}
require(qqman)

## First for Fe
fe_out = readRDS("./qtl/Fe/out.rds")
fe_out_df = data.frame("snp" = row.names(fe_out), "lod" = fe_out, row.names = NULL)
gwas_gmap = read.csv("./qtl/Fe/gwas_gmap.csv", stringsAsFactors = FALSE)
fe_for_manh = merge(fe_out_df, gwas_gmap, by.x = "snp", by.y = "marker")
fe_for_manh$chr[fe_for_manh$chr == "X"] = "20"
fe_for_manh = fe_for_manh[(fe_for_manh$chr != "M"),]
fe_for_manh$chr = as.numeric(fe_for_manh$chr)
fe_for_manh$pval = lodToPval(fe_for_manh$fpg)
write.csv(fe_for_manh, "./result/fe_snps.csv", row.names = FALSE)

pdf('./fe_manh.pdf', width = 8, height = 4)
p = manhattan(fe_for_manh, chr = "chr", bp = "pos", snp = "snp", p = "pval", logp = TRUE, ylim = c(0, 4), suggestiveline = F, genomewideline = F, main = "FPG, Fe", col = c("#003366", "#66B2FF"), cex = 0.5)
plot(p)
dev.off()

## Then for Ar
ar_out = readRDS("./qtl/Ar/out.rds")
ar_out_df = data.frame("snp" = row.names(ar_out), "lod" = ar_out, row.names = NULL)
gwas_gmap = read.csv("./qtl/Ar/gwas_gmap.csv", stringsAsFactors = FALSE)
ar_for_manh = merge(ar_out_df, gwas_gmap, by.x = "snp", by.y = "marker")
ar_for_manh$chr[ar_for_manh$chr == "X"] = "20"
ar_for_manh = ar_for_manh[(ar_for_manh$chr != "M"),]
ar_for_manh$chr = as.numeric(ar_for_manh$chr)
ar_for_manh$pval = lodToPval(ar_for_manh$fpg)
write.csv(ar_for_manh, "./result/ar_snps.csv", row.names = FALSE)

pdf('./ar_manh.pdf', width = 8, height = 4)
p = manhattan(ar_for_manh, chr = "chr", bp = "pos", snp = "snp", p = "pval", logp = TRUE, ylim = c(0, 4), suggestiveline = F, genomewideline = F, main = "FPG, Ar", col = c("#990000", "#FF9999"), cex = 0.5)
plot(p)
dev.off()

## Then for X-ray
xray_out = readRDS("./qtl/X-ray/out.rds")
xray_out_df = data.frame("snp" = row.names(xray_out), "lod" = xray_out, row.names = NULL)
gwas_gmap = read.csv("./qtl/X-ray/gwas_gmap.csv", stringsAsFactors = FALSE)
xray_for_manh = merge(xray_out_df, gwas_gmap, by.x = "snp", by.y = "marker")
xray_for_manh$chr[xray_for_manh$chr == "X"] = "20"
xray_for_manh = xray_for_manh[(xray_for_manh$chr != "M"),]
xray_for_manh$chr = as.numeric(xray_for_manh$chr)
xray_for_manh$pval = lodToPval(xray_for_manh$fpg)
write.csv(xray_for_manh, "./result/xray_snps.csv", row.names = FALSE)

pdf('./xray_manh.pdf', width = 8, height = 4)
p = manhattan(xray_for_manh, chr = "chr", bp = "pos", snp = "snp", p = "pval", logp = TRUE, ylim = c(0, 4), suggestiveline = F, genomewideline = F, main = "FPG, X-ray", col = c("#006600", "#66CC00"), cex = 0.5)
plot(p)
dev.off()

## Then for Background
bgd_out = readRDS("./qtl/Bgd/out.rds")
bgd_out_df = data.frame("snp" = row.names(bgd_out), "lod" = bgd_out, row.names = NULL)
gwas_gmap = read.csv("./qtl/Bgd/gwas_gmap.csv", stringsAsFactors = FALSE)
bgd_for_manh = merge(bgd_out_df, gwas_gmap, by.x = "snp", by.y = "marker")
bgd_for_manh$chr[bgd_for_manh$chr == "X"] = "20"
bgd_for_manh = bgd_for_manh[(bgd_for_manh$chr != "M"),]
bgd_for_manh$chr = as.numeric(bgd_for_manh$chr)
bgd_for_manh$pval = lodToPval(bgd_for_manh$fpg)
write.csv(bgd_for_manh, "./result/bgd_snps.csv", row.names = FALSE)

pdf('./bgd_manh.pdf', width = 8, height = 4)
p = manhattan(bgd_for_manh, chr = "chr", bp = "pos", snp = "snp", p = "pval", logp = TRUE, ylim = c(0, 4), suggestiveline = F, genomewideline = F, main = "Background", cex = 0.5)
plot(p)
dev.off()

```



```{r}
library(tidyverse)
library(fgsea)
all_res <- list()

for (rad in c("Ar", "Fe", "X-ray", "Bgd")) {
  print(rad)
  out <- readRDS(paste0("./qtl/", rad, "/out.rds"))
  for (combine_method in c("fisher", "addCLT")) {
    print(combine_method)
    
    snps_p <- rep(NA, nrow(snps))
    names(snps_p) <- snps$marker
    snps_p[intersect(snps$marker, rownames(out))] <- out[intersect(snps$marker, rownames(out)) ,1] # SNP p-values for condition
    snps_p <- lodToPval(snps_p)
    
    genes_p <- calculate_p(mapping, snps_p, F, combine_method) # Calculate p-value for each gene 
    genes_p <- genes_p[!is.na(genes_p)]
    genes_pfdr <- p.adjust(genes_p, "fdr")
    if(length(genes_p) == 0) next

    ora_res <- run_ORA(reactome_pathways$genesets, reactome_gene_set_names, genes_pfdr) # ORA result
    
    tmp <- genes_p
    tmp <- tmp*(1-2*1e-10) + 1e-10
    genes_z <- qnorm(tmp)
  
    fgsea_res <- run_FGSEA(reactome_pathways$genesets, reactome_gene_set_names, genes_z, genes_p) # FGSEA result
    
    # Result data frame
    res <- data.frame(pathway_name = reactome_gene_set_short_names, pathway = reactome_gene_set_names, ora_p = NA, ora_pfdr = NA, fgsea_p = NA, fgsea_pfdr = NA)
    res$ora_p[match(ora_res$pathway, reactome_gene_set_names)] <- ora_res$p.value
    res$ora_pfdr[match(ora_res$pathway, reactome_gene_set_names)] <- ora_res$p.adj
    res$fgsea_p[match(fgsea_res$pathway, reactome_gene_set_names)] <- fgsea_res$p.value
    res$fgsea_pfdr[match(fgsea_res$pathway, reactome_gene_set_names)] <- fgsea_res$p.adj
    
    con <- rad
    
    all_res[[con]] <- res

    if(!dir.exists(paste0("./result/",con,"/"))) dir.create(paste0("./result/",con,"/"), recursive = T)
    tmp <- res[, c(1,3,4)]
    tmp[is.na(tmp)] <- 1
    colnames(tmp) <- c("pathway", "p.value", "pFDR")
    write.csv(tmp, file = paste0("./result/",con,"/ORA_", combine_method,".csv"), row.names = F)
    tmp <- res[, c(1,5,6)]
    tmp[is.na(tmp)] <- 1
    colnames(tmp) <- c("pathway", "p.value", "pFDR")
    write.csv(tmp, file = paste0("./result/",con,"/FGSEA_", combine_method,".csv"), row.names = F)
      
    tmp <- data.frame(gene_id = intersect_genes, 
                     gene_symbol = intersect_genes_symbol,
                     p.value = genes_p[match(intersect_genes, as.numeric(names(genes_p)))],
                     pFDR = genes_pfdr[match(intersect_genes, as.numeric(names(genes_p)))])
    write.csv(tmp, file = paste0("./result/",con,"/gene_pvalue_", combine_method,".csv"), row.names = F)
  }
}


```

## Meta-analysis
```{r}
all_significant <- NULL
cons <- c("Ar", "Fe", "X-ray", "Bgd")
bgd_pathway <- NULL
bgd_con <- NULL
fpg_pathway <- NULL
fpg_con <- NULL
for (con in cons) {
  res <- NULL
  for (combine_method in c("fisher", "addCLT")) {
    if(!file.exists(paste0("./result/",con,"/ORA_", combine_method,".csv"))) next
    tmp <- read.csv(paste0("./result/",con,"/ORA_", combine_method,".csv"))
    tmp1 <- read.csv(paste0("./result/",con,"/FGSEA_", combine_method,".csv"))
    res <- cbind(res, tmp$p.value, tmp1$p.value)
    rownames(res) <- tmp$pathway
  }
  if(is.null(res)) next
  p <- apply(res, 1, addCLT) # Combine 4 p-values from 2 combining (Fisher and addCLT) and 2 pathway analysis (ORA and FGSEA) methods
  print(con)
  tmp <- data.frame(pathway = rownames(res), p.value = p, pFDR = p.adjust(p, "fdr"))
  tmp_withnames <- cbind(tmp, reactome_gene_set_names)
  colnames(tmp_withnames)[4] <- "pathway_name"
  write.csv(tmp_withnames, file = paste0("./result/",con,".csv"), row.names = F)

  tmp_withnames <- tmp_withnames[order(tmp_withnames[, 3]), c(4, 2, 3)]
  tmp_withnames$pathway_name <- as.character(tmp_withnames$pathway_name)
  tmp_withnames <- tmp_withnames[tmp_withnames[,3] < 0.05, ]
  rownames(tmp_withnames) <- NULL
  
  if(grepl("Bgd", con)) {
    bgd_pathway <- c(bgd_pathway, as.character(tmp_withnames$pathway_name))
    bgd_con <- c(bgd_con, rep(con, nrow(tmp_withnames)))
  } else {
    fpg_pathway <- c(fpg_pathway, as.character(tmp_withnames$pathway_name))
    fpg_con <- c(fpg_con, rep(con, nrow(tmp_withnames)))
  } 
  
  tmp_withnames <- rbind(c(con, "", ""), tmp_withnames)
  colnames(tmp_withnames) <- c("pathway_name", "p.value", "pFDR")
  all_significant <- rbind(all_significant, tmp_withnames)
}
write.csv(all_significant, file = paste0("./result/","all_significant.csv"), row.names = F)
```


```{r}
# Combine gene p-values from fisher and addCLT methods
all_significant <- NULL
cons <- c("Ar", "Fe", "X-ray", "Bgd")
all_genes_p <- list()
for (con in cons) {
  res <- NULL
  con <- stringr::str_replace(con, "/", "_")

  if(!file.exists(paste0("./result/",con,"/gene_pvalue_", "fisher",".csv"))) next
  tmp <- read.csv(paste0("./result/",con,"/gene_pvalue_", "fisher",".csv"))
  tmp1 <- read.csv(paste0("./result/",con,"/gene_pvalue_", "addCLT",".csv"))
  
  tmp <- cbind(tmp, tmp1[,c(3,4)])
  colnames(tmp) <- c("Entrez_ID", "Symbol", "fisher_pvalue", "fisher_pFDR", "addCLT_pvalue", "addCLT_pFDR")
  idx <- which(sapply(1:nrow(tmp), function(i) sum(is.na(tmp[i, 3:6])) < 4))
  tmp <- tmp[idx, ]

  write.csv(tmp, file = paste0("./result/gene_pvalue_",con,".csv"), row.names = F)
  all_genes_p[[con]] <- tmp

}
saveRDS(all_genes_p, "./result/all_genes_p.rds")
```

## Pathway histogram
```{r}
reactome_pathway_rel <- read.delim("./data/ReactomePathwaysRelation.txt", header = F)
reactome_pathway_rel <- reactome_pathway_rel[grepl("MMU", reactome_pathway_rel[,1]), ]

reactome_pathway_infor <- read.delim("./data/ReactomePathways.txt", header = F)
reactome_pathway_infor <- reactome_pathway_infor[grepl("MMU", reactome_pathway_infor[,1]), ]

top_pathway <- as.character(unique(reactome_pathway_rel[which(!reactome_pathway_rel[,1] %in% reactome_pathway_rel[,2]), 1]))

pathway_rel <- lapply(reactome_pathway_rel[,1], function(x) as.character(x) )
names(pathway_rel) <- as.character(reactome_pathway_rel[,2])
# Find which big group the pathway belongs to, then output the name of pathway
find_root <- function(pathway, pathway_rel, top_pathway, reactome_pathway_infor)
{
  while (!pathway %in% top_pathway) {
    pathway <- pathway_rel[[pathway]]
  }
  as.character(reactome_pathway_infor[which(reactome_pathway_infor$V1 == pathway), 2])
}

find_short_name <- function(pathway_name, reactome_pathway_infor)
{
  if (pathway_name %in% reactome_pathway_infor$V2) {
    return(reactome_pathway_infor$V1[which(reactome_pathway_infor$V2 == pathway_name)])
  } else {
    sim <- sapply(reactome_pathway_infor$V2, function(x) adist(pathway_name, x))
    return(reactome_pathway_infor$V1[which.min(sim)])
  }
}

fpg_pathway_short <- as.character(sapply(fpg_pathway, function(x) find_short_name(x, reactome_pathway_infor)) )

bgd_pathway_short <- as.character(sapply(bgd_pathway, function(x) find_short_name(x, reactome_pathway_infor)) ) ## EC addition

fpg_root <- sapply(fpg_pathway_short, function(x) find_root(x, pathway_rel, top_pathway, reactome_pathway_infor))

bgd_root <- sapply(bgd_pathway_short, function(x) find_root(x, pathway_rel, top_pathway, reactome_pathway_infor)) ## EC addition

library(ggsci)
library(ggplot2)
library(RColorBrewer)

tmp <- data.frame(x = fpg_root, color = fpg_con)
tmp$x <- factor(tmp$x, levels = names(sort(table(tmp$x), decreasing = T)) )
tmp$color <- factor(tmp$color, levels = unique(tmp$color))
levels(tmp$color) <- c("Ar", "Fe", "X-ray")

pdf('./fpg_hist.pdf', width = 8, height = 4)
p <- ggplot(tmp, aes(x = x, fill = color)) + geom_bar() + theme_classic() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + labs(x = NULL, y = "Number of pathways", fill="Condition") +
  theme(plot.margin = margin(0.5, 0, 0, 1.5, "cm")) + coord_cartesian(ylim = c(0, 40)) + scale_fill_npg()
plot(p)
dev.off()

## And the same for Bgd - EC addition

tmp <- data.frame(x = bgd_root, color = bgd_con)
tmp$x <- factor(tmp$x, levels = names(sort(table(tmp$x), decreasing = T)) )
tmp$color <- factor(tmp$color, levels = unique(tmp$color))
levels(tmp$color) <- c("Bgd")

pdf('./bgd_hist.pdf', width = 8, height = 4)
p <- ggplot(tmp, aes(x = x, fill = color)) + geom_bar() + theme_classic() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + labs(x = NULL, y = "Number of pathways", fill="Condition") +
  theme(plot.margin = margin(0.5, 0, 0, 1.5, "cm")) + coord_cartesian(ylim = c(0, 40)) + scale_fill_npg()
plot(p)
dev.off()

```

## Gene to SNP mapping
```{r}
gene_2_snp <- list()

for (gene in names(mapping)) {
  gene_2_snp[[gene]] <- NULL
  for (snp in mapping[[gene]]) {
    gene_2_snp[[gene]] <- c(gene_2_snp[[gene]], paste0(snps$chromosome[snp], ":", snps$position.b38.[snp]))
  }
}

saveRDS(gene_2_snp, "./result/gene_2_snp.rds")
```

## Hairball diagram for significant pathways
```{r}
library(ggraph)
library(igraph)
library(ggsci)

cons <- c("Ar", "Fe", "X-ray")
all_res <- list()
for (con in cons) {
  con <- stringr::str_replace(con, "/", "_")
  file_name <- paste0("./result/",con,".csv")
  if(!file.exists(file_name)) next
  tmp <- read.csv(file = file_name)
  tmp <- tmp[tmp$pFDR <= 0.05, ]
  all_res[[con]] <- tmp
}



vertex_name <- unique(unlist(lapply(c("Ar", "Fe", "X-ray"), function(rad) paste0(as.character(all_res[[rad]]$pathway), "_", rad) )))
vertex_rad <- unlist(lapply(c("Ar", "Fe", "X-ray"), function(rad) rep(rad, length(unique(all_res[[rad]]$pathway))) ))
edge_df <- data.frame(from = NULL, to = NULL, rad = NULL)

all_intersect <- intersect(intersect(all_res[["Ar"]]$pathway, all_res[["Fe"]]$pathway), all_res[["X-ray"]]$pathway)

combi <- combn(c("Ar", "Fe", "X-ray"), 2)
for (i in 1:ncol(combi)) {
  tmp <- all_intersect
  # tmp <- data.frame(paste0(tmp, "_", combi[1,i]), paste0(tmp, "_", combi[2,i]))
  tmp <- expand.grid(paste0(tmp, "_", combi[1,i]), paste0(tmp, "_", combi[2,i]))
  tmp <- cbind(tmp, "Ar - Fe - X-ray")
  colnames(tmp) <- c("from", "to", "rad")
  edge_df <- rbind(edge_df, tmp)
  
  tmp <- intersect(all_res[[combi[1,i]]]$pathway, all_res[[combi[2,i]]]$pathway)
  tmp <- tmp[!tmp %in% all_intersect]
  if(length(tmp) == 0) next
  # tmp <- data.frame(paste0(tmp, "_", combi[1,i]), paste0(tmp, "_", combi[2,i]))
  tmp <- expand.grid(paste0(tmp, "_", combi[1,i]), paste0(tmp, "_", combi[2,i]))
  tmp <- cbind(tmp, paste0(combi[1,i], " - ", combi[2,i]))
  colnames(tmp) <- c("from", "to", "rad")
  edge_df <- rbind(edge_df, tmp)
}

g <- graph_from_data_frame(edge_df, vertices = vertex_name)
from <- match(edge_df$from, vertex_name)
to <- match(edge_df$to, vertex_name)

pdf("hairball_2.pdf", height = 8, width = 8)
ggraph(g, layout = 'linear', circular = TRUE) + 
  geom_edge_arc(aes(edge_colour = edge_df$rad)) + scale_edge_colour_manual(values = pal_npg()(5)) +
  geom_node_point(aes(colour = vertex_rad)) + scale_color_aaas() +
  labs(colour = "Radiation", edge_colour = "Co-appearance") +
  coord_fixed() + theme(legend.key = element_rect(fill = "white", colour = NA), panel.background = element_rect(fill = "white", colour = NA))
dev.off()

## EC will try to make a hairball with nicer colors
pdf("hairball_2_updated.pdf", height = 8, width = 8)
ggraph(g, layout = 'linear', circular = TRUE) + 
  geom_edge_arc(aes(edge_colour = edge_df$rad)) + scale_edge_colour_manual(values = c("#800080", "#808080", "#F39B7FFF", "#008080")) +
  geom_node_point(aes(colour = vertex_rad)) + scale_colour_manual(values = c("#800000", "#000080", "#008000")) +
  labs(colour = "Radiation", edge_colour = "Co-appearance") +
  coord_fixed() + theme(legend.key = element_rect(fill = "white", colour = NA), panel.background = element_rect(fill = "white", colour = NA))
dev.off()

```

```{r}
cons <- c("Ar", "Fe", "X-ray")
all_res <- list()
for (con in cons) {
  con <- stringr::str_replace(con, "/", "_")
  file_name <- paste0("./result/",con,".csv")
  if(!file.exists(file_name)) next
  tmp <- read.csv(file = file_name)
  tmp <- tmp[tmp$pFDR <= 0.05, ]
  all_res[[con]] <- tmp
}

vertex_name <- unique(unlist(lapply(c("Ar", "Fe", "X-ray"), function(rad) paste0(as.character(all_res[[rad]]$pathway), "_", rad) )))
vertex_name1 <- unlist(lapply(c("Ar", "Fe", "X-ray"), function(rad) unique(as.character(all_res[[rad]]$pathway)) ))
vertex_rad <- unlist(lapply(c("Ar", "Fe", "X-ray"), function(rad) rep(rad, length(unique(all_res[[rad]]$pathway))) ))
vertex_name_long <- sapply(1:length(vertex_name1), function(i) as.character(all_res[[vertex_rad[i]]]$pathway_name[which(all_res[[vertex_rad[i]]]$pathway == vertex_name1[i])])[1] )

write.csv(data.frame(pathway = vertex_name1, pathway_name = vertex_name_long, radiation = vertex_rad), "./result/hairball_pathway.csv", row.names = F)
```

## Upset diagram for significant pathways
```{r}
library(UpSetR)


cons <- c("Ar", "Fe", "X-ray")
all_res <- list()
for (con in cons) {
  con <- stringr::str_replace(con, "/", "_")
  file_name <- paste0("./result/",con,".csv")
  if(!file.exists(file_name)) next
  tmp <- read.csv(file = file_name)
  tmp <- tmp[tmp$pFDR <= 0.05, ]
  all_res[[con]] <- tmp
}

sig_pathways <- lapply(c("Ar", "Fe", "X-ray"), function(rad) unique(as.character(all_res[[rad]]$pathway)) )
names(sig_pathways) <- c("Ar", "Fe", "X-ray")

pdf("upset_pathway.pdf", height = 6, width = 8)
upset(fromList(sig_pathways), order.by = "freq")
dev.off()
```


```{r}
library(UpSetR)

cons <- c("Ar", "Fe", "X-ray")
all_res <- list()
for (con in cons) {
  file_name <- paste0("./result/",con,".csv")
  if(!file.exists(file_name)) next
  tmp <- read.csv(file = file_name)
  tmp <- tmp[tmp$pFDR <= 0.05, ]
  all_res[[con]] <- tmp
}


vertex_name <- unique(unlist(lapply(c("Ar", "Fe", "X-ray"), function(rad) paste0(as.character(all_res[[rad]]$pathway), "_", rad) )))
vertex_name1 <- unlist(lapply(c("Ar", "Fe", "X-ray"), function(rad) unique(as.character(all_res[[rad]]$pathway)) ))
vertex_rad <- unlist(lapply(c("Ar", "Fe", "X-ray"), function(rad) rep(rad, length(unique(all_res[[rad]]$pathway))) ))
vertex_name_long <- sapply(1:length(vertex_name1), function(i) as.character(all_res[[vertex_rad[i]]]$pathway_name[which(all_res[[vertex_rad[i]]]$pathway == vertex_name1[i])])[1] )


cons <- c("Ar", "Fe", "X-ray")
all_res <- list()
for (con in cons) {
  file_name <- paste0("./result/",con,".csv")
  if(!file.exists(file_name)) next
  tmp <- read.csv(file = file_name)
  tmp <- tmp[tmp$pFDR <= 0.05, ]
  all_res[[con]] <- tmp
}


sig_pathways <- lapply(c("Ar", "Fe", "X-ray"), function(rad) unique(as.character(all_res[[rad]]$pathway)) )
names(sig_pathways) <- c("Ar", "Fe", "X-ray")

data <- matrix(0, ncol = 3, nrow = length(unique(unlist(sig_pathways))))
rownames(data) <- unique(unlist(sig_pathways))
colnames(data) <- c("Ar", "Fe", "X-ray")

for (rad in c("Ar", "Fe", "X-ray")) {
  data[sig_pathways[[rad]], rad] <- 1
}

subsets <- list(c("Ar"), c("Fe"), c("X-ray"), c("Ar", "Fe"), c("Ar", "X-ray"), c("Fe", "X-ray"), c("Ar", "Fe", "X-ray"))
tmp <- lapply(subsets, function(set) rownames(data)[which(rowSums(data[, set, drop=F]) == length(set) & rowSums(data) == length(set))])
names(tmp) <- sapply(subsets, function(x) paste0(x, collapse = " - "))

tmp <- lapply(tmp, function(x) data.frame(pathway = x, pathway_name = vertex_name_long[match(x, vertex_name1)], stringsAsFactors = F))

saveRDS(tmp, "./result/upset_pathway_group.rds")
```

## Upset diagram for significant genes
```{r}
library(UpSetR)


cons <- c("Ar", "Fe", "X-ray")
all_res <- list()
for (con in cons) {
  file_name <- paste0("./result/gene_pvalue_",con,".csv")
  if(!file.exists(file_name)) next
  tmp <- read.csv(file = file_name)
  tmp <- tmp[tmp$fisher_pFDR <= 0.05 & tmp$addCLT_pFDR <= 0.05, ]
  all_res[[con]] <- tmp
}

sig_genes <- lapply(c("Ar", "Fe", "X-ray"), function(rad) unique(as.character(all_res[[rad]]$Symbol)) )
names(sig_genes) <- c("Ar", "Fe", "X-ray")

pdf("upset_gene.pdf", height = 6, width = 8)
upset(fromList(sig_genes), order.by = "freq")
dev.off()
```

```{r}
## EC: overlapping genes for Venn diagrams
genes_fe = read.csv("./result/gene_pvalue_Fe.csv", stringsAsFactors = FALSE)
genes_ar = read.csv("./result/gene_pvalue_Ar.csv", stringsAsFactors = FALSE) 
genes_x = read.csv("./result/gene_pvalue_X-ray.csv", stringsAsFactors = FALSE)
fe_sig = genes_fe[(genes_fe$fisher_pFDR < 0.05 & genes_fe$addCLT_pFDR < 0.05),]
ar_sig = genes_ar[(genes_ar$fisher_pFDR < 0.05 & genes_ar$addCLT_pFDR < 0.05),]
x_sig = genes_x[(genes_x$fisher_pFDR < 0.05 & genes_x$addCLT_pFDR < 0.05),]
ar_fe = ar_sig[(ar_sig$Entrez_ID %in% fe_sig$Entrez_ID),]
ar_x = ar_sig[(ar_sig$Entrez_ID %in% x_sig$Entrez_ID),]
fe_x = fe_sig[(fe_sig$Entrez_ID %in% x_sig$Entrez_ID),]
ar_fe_x = fe_x[(fe_x$Entrez_ID %in% ar_sig$Entrez_ID),]
write.csv(ar_fe, "./result/sig_ar_fe.csv", row.names = FALSE)
write.csv(ar_x, "./result/sig_ar_xray.csv", row.names = FALSE)
write.csv(fe_x, "./result/sig_fe_xray.csv", row.names = FALSE)
write.csv(ar_fe_x, "./result/sig_ar_fe_xray.csv", row.names = FALSE)
```


